File: python/src/hron/_lexer.py
from __future__ import annotations

from dataclasses import dataclass

from ._ast import IntervalUnit, MonthName, OrdinalPosition, Weekday
from ._error import HronError, Span

# --- Token kinds ---


@dataclass(frozen=True, slots=True)
class TEvery:
    pass


@dataclass(frozen=True, slots=True)
class TOn:
    pass


@dataclass(frozen=True, slots=True)
class TAt:
    pass


@dataclass(frozen=True, slots=True)
class TFrom:
    pass


@dataclass(frozen=True, slots=True)
class TTo:
    pass


@dataclass(frozen=True, slots=True)
class TIn:
    pass


@dataclass(frozen=True, slots=True)
class TOf:
    pass


@dataclass(frozen=True, slots=True)
class TThe:
    pass


@dataclass(frozen=True, slots=True)
class TLast:
    pass


@dataclass(frozen=True, slots=True)
class TExcept:
    pass


@dataclass(frozen=True, slots=True)
class TUntil:
    pass


@dataclass(frozen=True, slots=True)
class TStarting:
    pass


@dataclass(frozen=True, slots=True)
class TDuring:
    pass


@dataclass(frozen=True, slots=True)
class TNearest:
    pass


@dataclass(frozen=True, slots=True)
class TNext:
    pass


@dataclass(frozen=True, slots=True)
class TPrevious:
    pass


@dataclass(frozen=True, slots=True)
class TYear:
    pass


@dataclass(frozen=True, slots=True)
class TDay:
    pass


@dataclass(frozen=True, slots=True)
class TWeekday:
    pass


@dataclass(frozen=True, slots=True)
class TWeekend:
    pass


@dataclass(frozen=True, slots=True)
class TWeeks:
    pass


@dataclass(frozen=True, slots=True)
class TMonth:
    pass


@dataclass(frozen=True, slots=True)
class TDayName:
    name: Weekday


@dataclass(frozen=True, slots=True)
class TMonthName:
    name: MonthName


@dataclass(frozen=True, slots=True)
class TOrdinal:
    name: OrdinalPosition


@dataclass(frozen=True, slots=True)
class TIntervalUnit:
    unit: IntervalUnit


@dataclass(frozen=True, slots=True)
class TNumber:
    value: int


@dataclass(frozen=True, slots=True)
class TOrdinalNumber:
    value: int


@dataclass(frozen=True, slots=True)
class TTime:
    hour: int
    minute: int


@dataclass(frozen=True, slots=True)
class TIsoDate:
    date: str


@dataclass(frozen=True, slots=True)
class TComma:
    pass


@dataclass(frozen=True, slots=True)
class TTimezone:
    tz: str


TokenKind = (
    TEvery
    | TOn
    | TAt
    | TFrom
    | TTo
    | TIn
    | TOf
    | TThe
    | TLast
    | TExcept
    | TUntil
    | TStarting
    | TDuring
    | TNearest
    | TNext
    | TPrevious
    | TYear
    | TDay
    | TWeekday
    | TWeekend
    | TWeeks
    | TMonth
    | TDayName
    | TMonthName
    | TOrdinal
    | TIntervalUnit
    | TNumber
    | TOrdinalNumber
    | TTime
    | TIsoDate
    | TComma
    | TTimezone
)


@dataclass(frozen=True, slots=True)
class Token:
    kind: TokenKind
    span: Span


# --- Keyword map ---

_KEYWORD_MAP: dict[str, TokenKind] = {
    "every": TEvery(),
    "on": TOn(),
    "at": TAt(),
    "from": TFrom(),
    "to": TTo(),
    "in": TIn(),
    "of": TOf(),
    "the": TThe(),
    "last": TLast(),
    "except": TExcept(),
    "until": TUntil(),
    "starting": TStarting(),
    "during": TDuring(),
    "nearest": TNearest(),
    "next": TNext(),
    "previous": TPrevious(),
    "year": TYear(),
    "years": TYear(),
    "day": TDay(),
    "days": TDay(),
    "weekday": TWeekday(),
    "weekdays": TWeekday(),
    "weekend": TWeekend(),
    "weekends": TWeekend(),
    "weeks": TWeeks(),
    "week": TWeeks(),
    "month": TMonth(),
    "months": TMonth(),
    # Day names
    "monday": TDayName(Weekday.MONDAY),
    "mon": TDayName(Weekday.MONDAY),
    "tuesday": TDayName(Weekday.TUESDAY),
    "tue": TDayName(Weekday.TUESDAY),
    "wednesday": TDayName(Weekday.WEDNESDAY),
    "wed": TDayName(Weekday.WEDNESDAY),
    "thursday": TDayName(Weekday.THURSDAY),
    "thu": TDayName(Weekday.THURSDAY),
    "friday": TDayName(Weekday.FRIDAY),
    "fri": TDayName(Weekday.FRIDAY),
    "saturday": TDayName(Weekday.SATURDAY),
    "sat": TDayName(Weekday.SATURDAY),
    "sunday": TDayName(Weekday.SUNDAY),
    "sun": TDayName(Weekday.SUNDAY),
    # Month names
    "january": TMonthName(MonthName.JAN),
    "jan": TMonthName(MonthName.JAN),
    "february": TMonthName(MonthName.FEB),
    "feb": TMonthName(MonthName.FEB),
    "march": TMonthName(MonthName.MAR),
    "mar": TMonthName(MonthName.MAR),
    "april": TMonthName(MonthName.APR),
    "apr": TMonthName(MonthName.APR),
    "may": TMonthName(MonthName.MAY),
    "june": TMonthName(MonthName.JUN),
    "jun": TMonthName(MonthName.JUN),
    "july": TMonthName(MonthName.JUL),
    "jul": TMonthName(MonthName.JUL),
    "august": TMonthName(MonthName.AUG),
    "aug": TMonthName(MonthName.AUG),
    "september": TMonthName(MonthName.SEP),
    "sep": TMonthName(MonthName.SEP),
    "october": TMonthName(MonthName.OCT),
    "oct": TMonthName(MonthName.OCT),
    "november": TMonthName(MonthName.NOV),
    "nov": TMonthName(MonthName.NOV),
    "december": TMonthName(MonthName.DEC),
    "dec": TMonthName(MonthName.DEC),
    # Ordinals
    "first": TOrdinal(OrdinalPosition.FIRST),
    "second": TOrdinal(OrdinalPosition.SECOND),
    "third": TOrdinal(OrdinalPosition.THIRD),
    "fourth": TOrdinal(OrdinalPosition.FOURTH),
    "fifth": TOrdinal(OrdinalPosition.FIFTH),
    # Interval units
    "min": TIntervalUnit(IntervalUnit.MIN),
    "mins": TIntervalUnit(IntervalUnit.MIN),
    "minute": TIntervalUnit(IntervalUnit.MIN),
    "minutes": TIntervalUnit(IntervalUnit.MIN),
    "hour": TIntervalUnit(IntervalUnit.HOURS),
    "hours": TIntervalUnit(IntervalUnit.HOURS),
    "hr": TIntervalUnit(IntervalUnit.HOURS),
    "hrs": TIntervalUnit(IntervalUnit.HOURS),
}


class _Lexer:
    def __init__(self, input_text: str) -> None:
        self._input = input_text
        self._pos = 0
        self._after_in = False

    def tokenize(self) -> list[Token]:
        tokens: list[Token] = []
        while True:
            self._skip_whitespace()
            if self._pos >= len(self._input):
                break

            if self._after_in:
                self._after_in = False
                tokens.append(self._lex_timezone())
                continue

            start = self._pos
            ch = self._input[self._pos]

            if ch == ",":
                self._pos += 1
                tokens.append(Token(TComma(), Span(start, self._pos)))
                continue

            if ch.isdigit():
                tokens.append(self._lex_number_or_time_or_date())
                continue

            if ch.isalpha():
                tokens.append(self._lex_word())
                continue

            raise HronError.lex(
                f"unexpected character '{ch}'",
                Span(start, start + 1),
                self._input,
            )

        return tokens

    def _skip_whitespace(self) -> None:
        while self._pos < len(self._input) and self._input[self._pos] in " \t\n\r":
            self._pos += 1

    def _lex_timezone(self) -> Token:
        self._skip_whitespace()
        start = self._pos
        while self._pos < len(self._input) and self._input[self._pos] not in " \t\n\r":
            self._pos += 1
        tz = self._input[start : self._pos]
        if len(tz) == 0:
            raise HronError.lex(
                "expected timezone after 'in'",
                Span(start, start + 1),
                self._input,
            )
        return Token(TTimezone(tz), Span(start, self._pos))

    def _lex_number_or_time_or_date(self) -> Token:
        start = self._pos
        num_start = self._pos
        while self._pos < len(self._input) and self._input[self._pos].isdigit():
            self._pos += 1
        digits = self._input[num_start : self._pos]

        # Check for ISO date: YYYY-MM-DD
        if len(digits) == 4 and self._pos < len(self._input) and self._input[self._pos] == "-":
            remaining = self._input[start:]
            if (
                len(remaining) >= 10
                and remaining[4] == "-"
                and remaining[5].isdigit()
                and remaining[6].isdigit()
                and remaining[7] == "-"
                and remaining[8].isdigit()
                and remaining[9].isdigit()
            ):
                self._pos = start + 10
                return Token(TIsoDate(self._input[start : self._pos]), Span(start, self._pos))

        # Check for time: HH:MM
        if len(digits) in (1, 2) and self._pos < len(self._input) and self._input[self._pos] == ":":
            self._pos += 1  # skip ':'
            min_start = self._pos
            while self._pos < len(self._input) and self._input[self._pos].isdigit():
                self._pos += 1
            min_digits = self._input[min_start : self._pos]
            if len(min_digits) == 2:
                hour = int(digits)
                minute = int(min_digits)
                if hour > 23 or minute > 59:
                    raise HronError.lex("invalid time", Span(start, self._pos), self._input)
                return Token(TTime(hour, minute), Span(start, self._pos))

        num = int(digits)

        # Check for ordinal suffix: st, nd, rd, th
        if self._pos + 1 < len(self._input):
            suffix = self._input[self._pos : self._pos + 2].lower()
            if suffix in ("st", "nd", "rd", "th"):
                self._pos += 2
                return Token(TOrdinalNumber(num), Span(start, self._pos))

        return Token(TNumber(num), Span(start, self._pos))

    def _lex_word(self) -> Token:
        start = self._pos
        while self._pos < len(self._input) and (
            self._input[self._pos].isalnum() or self._input[self._pos] == "_"
        ):
            self._pos += 1
        word = self._input[start : self._pos].lower()
        span = Span(start, self._pos)

        kind = _KEYWORD_MAP.get(word)
        if kind is None:
            raise HronError.lex(f"unknown keyword '{word}'", span, self._input)

        if isinstance(kind, TIn):
            self._after_in = True

        return Token(kind, span)


def tokenize(input_text: str) -> list[Token]:
    return _Lexer(input_text).tokenize()

File: ruby/lib/hron/lexer.rb
# frozen_string_literal: true

require_relative "ast"
require_relative "error"

module Hron
  # Token kinds (using symbols and Data classes)
  module TokenKind
    EVERY = :every
    ON = :on
    AT = :at
    FROM = :from
    TO = :to
    IN = :in
    OF = :of
    THE = :the
    LAST = :last
    EXCEPT = :except
    UNTIL = :until
    STARTING = :starting
    DURING = :during
    YEAR = :year
    NEAREST = :nearest
    NEXT = :next
    PREVIOUS = :previous
    DAY = :day
    WEEKDAY_KW = :weekday_kw
    WEEKEND_KW = :weekend_kw
    WEEKS = :weeks
    MONTH = :month
    COMMA = :comma
  end

  # Token types with values
  TDayName = Data.define(:name)
  TMonthName = Data.define(:name)
  TOrdinal = Data.define(:position)
  TIntervalUnit = Data.define(:unit)
  TNumber = Data.define(:value)
  TOrdinalNumber = Data.define(:value)
  TTime = Data.define(:hour, :minute)
  TIsoDate = Data.define(:date)
  TTimezone = Data.define(:tz)

  # Token with kind and span
  Token = Data.define(:kind, :span)

  # Keyword mapping
  KEYWORD_MAP = {
    "every" => TokenKind::EVERY,
    "on" => TokenKind::ON,
    "at" => TokenKind::AT,
    "from" => TokenKind::FROM,
    "to" => TokenKind::TO,
    "in" => TokenKind::IN,
    "of" => TokenKind::OF,
    "the" => TokenKind::THE,
    "last" => TokenKind::LAST,
    "except" => TokenKind::EXCEPT,
    "until" => TokenKind::UNTIL,
    "starting" => TokenKind::STARTING,
    "during" => TokenKind::DURING,
    "year" => TokenKind::YEAR,
    "years" => TokenKind::YEAR,
    "nearest" => TokenKind::NEAREST,
    "next" => TokenKind::NEXT,
    "previous" => TokenKind::PREVIOUS,
    "day" => TokenKind::DAY,
    "days" => TokenKind::DAY,
    "weekday" => TokenKind::WEEKDAY_KW,
    "weekdays" => TokenKind::WEEKDAY_KW,
    "weekend" => TokenKind::WEEKEND_KW,
    "weekends" => TokenKind::WEEKEND_KW,
    "weeks" => TokenKind::WEEKS,
    "week" => TokenKind::WEEKS,
    "month" => TokenKind::MONTH,
    "months" => TokenKind::MONTH,
    # Day names
    "monday" => TDayName.new(Weekday::MONDAY),
    "mon" => TDayName.new(Weekday::MONDAY),
    "tuesday" => TDayName.new(Weekday::TUESDAY),
    "tue" => TDayName.new(Weekday::TUESDAY),
    "wednesday" => TDayName.new(Weekday::WEDNESDAY),
    "wed" => TDayName.new(Weekday::WEDNESDAY),
    "thursday" => TDayName.new(Weekday::THURSDAY),
    "thu" => TDayName.new(Weekday::THURSDAY),
    "friday" => TDayName.new(Weekday::FRIDAY),
    "fri" => TDayName.new(Weekday::FRIDAY),
    "saturday" => TDayName.new(Weekday::SATURDAY),
    "sat" => TDayName.new(Weekday::SATURDAY),
    "sunday" => TDayName.new(Weekday::SUNDAY),
    "sun" => TDayName.new(Weekday::SUNDAY),
    # Month names
    "january" => TMonthName.new(MonthName::JAN),
    "jan" => TMonthName.new(MonthName::JAN),
    "february" => TMonthName.new(MonthName::FEB),
    "feb" => TMonthName.new(MonthName::FEB),
    "march" => TMonthName.new(MonthName::MAR),
    "mar" => TMonthName.new(MonthName::MAR),
    "april" => TMonthName.new(MonthName::APR),
    "apr" => TMonthName.new(MonthName::APR),
    "may" => TMonthName.new(MonthName::MAY),
    "june" => TMonthName.new(MonthName::JUN),
    "jun" => TMonthName.new(MonthName::JUN),
    "july" => TMonthName.new(MonthName::JUL),
    "jul" => TMonthName.new(MonthName::JUL),
    "august" => TMonthName.new(MonthName::AUG),
    "aug" => TMonthName.new(MonthName::AUG),
    "september" => TMonthName.new(MonthName::SEP),
    "sep" => TMonthName.new(MonthName::SEP),
    "october" => TMonthName.new(MonthName::OCT),
    "oct" => TMonthName.new(MonthName::OCT),
    "november" => TMonthName.new(MonthName::NOV),
    "nov" => TMonthName.new(MonthName::NOV),
    "december" => TMonthName.new(MonthName::DEC),
    "dec" => TMonthName.new(MonthName::DEC),
    # Ordinals
    "first" => TOrdinal.new(OrdinalPosition::FIRST),
    "second" => TOrdinal.new(OrdinalPosition::SECOND),
    "third" => TOrdinal.new(OrdinalPosition::THIRD),
    "fourth" => TOrdinal.new(OrdinalPosition::FOURTH),
    "fifth" => TOrdinal.new(OrdinalPosition::FIFTH),
    # Interval units
    "min" => TIntervalUnit.new(IntervalUnit::MIN),
    "mins" => TIntervalUnit.new(IntervalUnit::MIN),
    "minute" => TIntervalUnit.new(IntervalUnit::MIN),
    "minutes" => TIntervalUnit.new(IntervalUnit::MIN),
    "hour" => TIntervalUnit.new(IntervalUnit::HOURS),
    "hours" => TIntervalUnit.new(IntervalUnit::HOURS),
    "hr" => TIntervalUnit.new(IntervalUnit::HOURS),
    "hrs" => TIntervalUnit.new(IntervalUnit::HOURS)
  }.freeze

  # Lexer class
  class Lexer
    def initialize(input)
      @input = input
      @pos = 0
      @after_in = false
    end

    def tokenize
      tokens = []
      loop do
        skip_whitespace
        break if @pos >= @input.length

        if @after_in
          @after_in = false
          tokens << lex_timezone
          next
        end

        start = @pos
        ch = @input[@pos]

        if ch == ","
          @pos += 1
          tokens << Token.new(TokenKind::COMMA, Span.new(start, @pos))
          next
        end

        if ch.match?(/\d/)
          tokens << lex_number_or_time_or_date
          next
        end

        if ch.match?(/[a-zA-Z]/)
          tokens << lex_word
          next
        end

        raise HronError.lex("unexpected character '#{ch}'", Span.new(start, start + 1), @input)
      end
      tokens
    end

    private

    def skip_whitespace
      @pos += 1 while @pos < @input.length && @input[@pos].match?(/\s/)
    end

    def lex_timezone
      skip_whitespace
      start = @pos
      @pos += 1 while @pos < @input.length && !@input[@pos].match?(/\s/)
      tz = @input[start...@pos]
      raise HronError.lex("expected timezone after 'in'", Span.new(start, start + 1), @input) if tz.empty?

      Token.new(TTimezone.new(tz), Span.new(start, @pos))
    end

    def lex_number_or_time_or_date
      start = @pos
      @pos += 1 while @pos < @input.length && @input[@pos].match?(/\d/)
      digits = @input[start...@pos]

      # Check for ISO date: YYYY-MM-DD
      if digits.length == 4 && @pos < @input.length && @input[@pos] == "-"
        remaining = @input[start..]
        if remaining.length >= 10 &&
            remaining[4] == "-" &&
            remaining[5..6].match?(/\d{2}/) &&
            remaining[7] == "-" &&
            remaining[8..9].match?(/\d{2}/)
          @pos = start + 10
          return Token.new(TIsoDate.new(@input[start...@pos]), Span.new(start, @pos))
        end
      end

      # Check for time: HH:MM
      if digits.length.between?(1, 2) && @pos < @input.length && @input[@pos] == ":"
        @pos += 1 # skip ':'
        min_start = @pos
        @pos += 1 while @pos < @input.length && @input[@pos].match?(/\d/)
        min_digits = @input[min_start...@pos]
        if min_digits.length == 2
          hour = digits.to_i
          minute = min_digits.to_i
          raise HronError.lex("invalid time", Span.new(start, @pos), @input) if hour > 23 || minute > 59

          return Token.new(TTime.new(hour, minute), Span.new(start, @pos))
        end
      end

      num = digits.to_i

      # Check for ordinal suffix: st, nd, rd, th
      if @pos + 1 < @input.length
        suffix = @input[@pos, 2].downcase
        if %w[st nd rd th].include?(suffix)
          @pos += 2
          return Token.new(TOrdinalNumber.new(num), Span.new(start, @pos))
        end
      end

      Token.new(TNumber.new(num), Span.new(start, @pos))
    end

    def lex_word
      start = @pos
      @pos += 1 while @pos < @input.length && @input[@pos].match?(/\w/)
      word = @input[start...@pos].downcase
      span = Span.new(start, @pos)

      kind = KEYWORD_MAP[word]
      raise HronError.lex("unknown keyword '#{word}'", span, @input) if kind.nil?

      @after_in = true if kind == TokenKind::IN

      Token.new(kind, span)
    end
  end

  def self.tokenize(input)
    Lexer.new(input).tokenize
  end
end
